Guide rewrite: z-index and stacking context (#38601)
mdn/content: files/en-us/web/css/css_positioned_layout/stacking_context/index.md

- Element that has had stacking context-creating properties (such as `opacity`) animated using {{cssxref("@keyframes")}}, with [`animation-fill-mode`](/en-US/docs/Web/CSS/animation-fill-mode) set to [`forwards`](/en-US/docs/Web/CSS/animation-fill-mode#forwards).

## Nested stacking contexts

Stacking contexts can be contained in other stacking contexts, and they can together create a hierarchy of stacking contexts.

The root element of a document is a stacking context which, in most cases, contains nested stacking contexts, many of which will contain additional stacking contexts. Within each stacking context, child elements are stacked according to the same rules explained in [Using `z-index`](/en-US/docs/Web/CSS/CSS_positioned_layout/Using_z-index). Importantly, the `z-index` values of its child stacking contexts only have meaning within its parent's stacking context. Stacking contexts are treated atomically as a single unit in the parent stacking context.

> To figure out the _rendering order_ of stacked elements along the z-axis think of each index value as a "version number" of sorts, where child elements are minor version numbers underneath their parent's major version numbers.

Reviewer: ```suggestion
To figure out the _rendering order_ of stacked elements along the z-axis, think of each index value as a "version number" of sorts, where child elements represent minor version numbers underneath their parent's major version number.
```

> To demonstrate how the stacking order of each element participates in the stacking order of their ancestor stacking contexts, let's look at an example page with six container elements. There are three sibling {{htmlelement("section")}} elements. The last `<section>` contains three sibling {{htmlelement("article")}} elements, with the {{htmlelement("heading_elements", "&lt;h1&gt;")}} and {{htmlelement("code")}} of that third section appearing between the after the first `<article>`.

Reviewer: - "ancestor's"
- the last sentence is still incomplete
- there is a mismatch between the text and code. From what I see in the code, the last `<article>` contains three sibling `<section>` elements but the text says "The last `<section>` contains three sibling {{htmlelement("article")}} elements"

> To demonstrate how the stacking order of each element participates in the stacking order of their ancestor stacking contexts, let's look at an example page with six container elements. There are three sibling {{htmlelement("section")}} elements. The last `<section>` contains three sibling {{htmlelement("article")}} elements, with the {{htmlelement("heading_elements", "&lt;h1&gt;")}} and {{htmlelement("code")}} of that third section appearing between the after the first `<article>`.

Author: ```suggestion
To demonstrate how the stacking order of each element participates in the stacking order of their ancestor stacking contexts, let's look at an example page with six container elements. There are three sibling {{htmlelement("article")}} elements. The last `<article>` contains three sibling {{htmlelement("section")}} elements, with the {{htmlelement("heading_elements", "&lt;h1&gt;")}} and {{htmlelement("code")}} of that third article appearing between the first and second sibling `<article>` elements.
```

- Element that has had stacking context-creating properties (such as `opacity`) animated using {{cssxref("@keyframes")}}, with [`animation-fill-mode`](/en-US/docs/Web/CSS/animation-fill-mode) set to [`forwards`](/en-US/docs/Web/CSS/animation-fill-mode#forwards).

## Nested stacking contexts

Stacking contexts can be contained in other stacking contexts, and together create a hierarchy of stacking contexts.

The root element of a document is a stacking context which in most cases contains nested stacking contexts, many of which will contain additional stacking contexts. Within each stacking context, child elements are stacked according to the same rules explained in [Using `z-index`](/en-US/docs/Web/CSS/CSS_positioned_layout/Using_z-index). Importantly, the `z-index` values of its child stacking contexts only have meaning within its parent's stacking context. Stacking contexts are treated atomically as a single unit in the parent stacking context.

> To figure out the _rendering order_ of stacked elements along the z-axis think of the it as a "version number" of sorts, where child elements are minor version numbers underneath their parent's major version numbers.

Reviewer: thanks

    12px/20px Arial,
    sans-serif;
}
h1 {
  font-size: 1.25em;
}
#container1,
#container2 {
  border: 1px dashed #696;
  padding: 10px;
  background-color: #cfc;
}
#container1 {
  margin-bottom: 190px;
}
#container3 {
  border: 1px dashed #900;
  background-color: #fdd;
  padding: 40px 20px 20px;
  width: 330px;
}
#container4 {
  border: 1px dashed #996;
  background-color: #ffc;
  padding: 25px 10px 5px;
  margin-bottom: 15px;
}
#container5 {
  border: 1px dashed #996;
  background-color: #ffc;
  margin-top: 15px;
  padding: 5px 10px;
}
#container6 {
  background-color: #ddf;
  border: 1px dashed #009;
  padding-left: 20px;
  padding-top: 125px;
  width: 150px;
  height: 125px;
}
```

```css
section,
article {
  opacity: 0.85;
  position: relative;
}
#container1 {
  z-index: 5;
}
#container2 {
  z-index: 2;
}
#container3 {
  z-index: 4;
  position: absolute;
  top: 40px;
  left: 180px;
}
#container4 {
  z-index: 6;
}
#container5 {
  z-index: 1;
}
#container6 {
  z-index: 3;
  position: absolute;
  top: 20px;
  left: 180px;
}
```

> The CSS for the colors, fonts, alignment, and [box-model](/en-US/docs/Web/CSS/CSS_box_model/Introduction_to_the_CSS_box_model) properties has been hidden for brevity.

Reviewer: ```suggestion
The CSS properties for colors, fonts, alignment, and [box-model](/en-US/docs/Web/CSS/CSS_box_model/Introduction_to_the_CSS_box_model) have been hidden for brevity.
```

{{ EmbedLiveSample('Nested stacking contexts', '100%', '396') }}

> The hierarchy of stacking contexts is organized as follows:
> 
> - Root
> 
>   - SECTION #1
>   - SECTION #2
>   - SECTION #3
> 
>     - ARTICLE #4
>     - ARTICLE #5
>     - ARTICLE #6

Reviewer: ```suggestion
The hierarchy of stacking contexts in the above example is as follows:

Root
│
├── SECTION #1
├── SECTION #2
└── SECTION #3
    │
    ├── ARTICLE #4
    ├── ARTICLE #5
    └── ARTICLE #6
```

>   └── ARTICLE #6

Reviewer: Do you want to revert to the bulleted list format you had previously?

The three articles are children of SECTION #3, so stacking of those elements is completely resolved within SECTION #3. Once stacking and rendering within SECTION #3 is completed, the whole SECTION #3 element is passed for stacking in the root element with respect to its sibling sections.

> ARTICLE #4 is rendered under SECTION #1 because SECTION #1's z-index (`5`) is valid within the stacking context of the root element, while ARTICLE #4's z-index (`6`) is valid within the stacking context of SECTION #3 (`z-index: 4`). So, ARTICLE #4 is under SECTION #1, because ARTICLE #4 belongs to SECTION #3, which has a lower z-index value (`4-6` is less than `5-0`).

Reviewer: ```suggestion
ARTICLE #4 is rendered under SECTION #1 because SECTION #1's z-index (`5`) is valid within the stacking context of the root element, while ARTICLE #4's z-index (`6`) is valid within the stacking context of SECTION #3 (`z-index: 4`). So ARTICLE #4 is under SECTION #1 because ARTICLE #4 belongs to SECTION #3, which has a lower z-index value (`4-6` is less than `5-0`).
```

> For the same reason SECTION #2 (`z-index: 2`) is rendered under ARTICLE #5 (`z-index`: 1) because ARTICLE #5 belongs to SECTION #3 (`z-index: 4`), which has a higher z-index value (`2-0` is less than `4-1`).

Reviewer: ```suggestion
For the same reason, SECTION #2 (`z-index: 2`) is rendered under ARTICLE #5 (`z-index`: 1) because ARTICLE #5 belongs to SECTION #3 (`z-index: 4`), which has a higher z-index value (`2-0` is less than `4-1`).
```

> SECTION #3's z-index is `4`, but this value is independent from z-index of three articles nested inside it, because it belongs to a different stacking context.

Reviewer: ```suggestion
SECTION #3's z-index is `4`, but this value is independent from z-index of three articles nested inside it because they belong to a different stacking context.
```

By comparing the z-index as "version numbers", we can see how an element with a z-index of 1 (ARTICLE #5) is stacked above an element with a z-index of 2 (SECTION #2), and how an element with a z-index of 6 (ARTICLE #4) is stacked below an element with a z-index of 5 (SECTION #1).

In our example (sorted according to the final rendering order):

- Root

  - SECTION #2: (`z-index`: 2), which results in a rendering order of `2-0`
  - SECTION #3: (`z-index`: 4), which results in a rendering order of `4-0`

    - ARTICLE #5: (`z-index`: 1), stacked under an element (`z-index`: 4), which results in a rendering order of `4-1`
    - ARTICLE #6: (`z-index`: 3), stacked under an element (`z-index`: 4), which results in a rendering order of `4-3`
    - ARTICLE #4: (`z-index`: 6), stacked under an element (`z-index`: 4), which results in a rendering order of `4-6`

>   - SECTION #1: (`z-index`: 5), which results in a rendering order of `5-0`

Reviewer: I suggest using the ASCII tree structure to better represent the hierarchy

Reviewer: Got it. In that case, do you want to undo the changes in https://github.com/mdn/content/pull/38601#discussion_r2002185616?

> ### Other examples

Reviewer: ```suggestion
## Additional examples
```

  background-color: #fdd;
  padding: 40px 20px 20px;
  width: 330px;
}
#container4 {
  border: 1px dashed #996;
  background-color: #ffc;
  padding: 25px 10px 5px;
  margin-bottom: 15px;
}
#container5 {
  border: 1px dashed #996;
  background-color: #ffc;
  margin-top: 15px;
  padding: 5px 10px;
}
#container6 {
  background-color: #ddf;
  border: 1px dashed #009;
  padding-left: 20px;
  padding-top: 125px;
  width: 150px;
  height: 125px;
}
```

```css
section,
article {
  opacity: 0.85;
  position: relative;
}
#container1 {
  z-index: 5;
}
#container2 {
  z-index: 2;
}
#container3 {
  z-index: 4;
  position: absolute;
  top: 40px;
  left: 180px;
}
#container4 {
  z-index: 6;
}
#container5 {
  z-index: 1;
}
#container6 {
  z-index: 3;
  position: absolute;
  top: 20px;
  left: 180px;
}
```

The CSS properties for colors, fonts, alignment, and [box-model](/en-US/docs/Web/CSS/CSS_box_model/Introduction_to_the_CSS_box_model) have been hidden for brevity.

{{ EmbedLiveSample('Nested stacking contexts', '100%', '396') }}

The hierarchy of stacking contexts in the above example is as follows:

```plain no-lint
Root
│
├── SECTION #1
├── SECTION #2
└── SECTION #3
  │

>   ├── ARTICLE #4
>   ├── ARTICLE #5

Author: ```suggestion
  ├── ARTICLE #4
  ├────  SECTION #3 content 
  ├── ARTICLE #5
```

}
#container4 {
  border: 1px dashed #996;
  background-color: #ffc;
  padding: 25px 10px 5px;
  margin-bottom: 15px;
}
#container5 {
  border: 1px dashed #996;
  background-color: #ffc;
  margin-top: 15px;
  padding: 5px 10px;
}
#container6 {
  background-color: #ddf;
  border: 1px dashed #009;
  padding-left: 20px;
  padding-top: 125px;
  width: 150px;
  height: 125px;
}
```

```css
section,
article {
  opacity: 0.85;
  position: relative;
}
#container1 {
  z-index: 5;
}
#container2 {
  z-index: 2;
}
#container3 {
  z-index: 4;
  position: absolute;
  top: 40px;
  left: 180px;
}
#container4 {
  z-index: 6;
}
#container5 {
  z-index: 1;
}
#container6 {
  z-index: 3;
  position: absolute;
  top: 20px;
  left: 180px;
}
```

The CSS properties for colors, fonts, alignment, and [box-model](/en-US/docs/Web/CSS/CSS_box_model/Introduction_to_the_CSS_box_model) have been hidden for brevity.

{{ EmbedLiveSample('Nested stacking contexts', '100%', '396') }}

The hierarchy of stacking contexts in the above example is as follows:

```plain no-lint
Root
│
├── SECTION #1
├── SECTION #2
└── SECTION #3
  │
  ├── ARTICLE #4
  ├────  SECTION #3 content
  ├── ARTICLE #5
  └── ARTICLE #6

> ```

Reviewer: shouldn't this be updated to show the 3 article elements at the root and section elements as nested?

SECTION #3's z-index is `4`, but this value is independent from z-index of three articles nested inside it because they belong to a different stacking context.

By comparing the z-index as "version numbers", we can see how an element with a z-index of 1 (ARTICLE #5) is stacked above an element with a z-index of 2 (SECTION #2), and how an element with a z-index of 6 (ARTICLE #4) is stacked below an element with a z-index of 5 (SECTION #1).

In our example (sorted according to the final rendering order):

- Root

  - SECTION #2: (`z-index`: 2), which results in a rendering order of `2-0`
  - SECTION #3: (`z-index`: 4), which results in a rendering order of `4-0`

    - ARTICLE #5: (`z-index`: 1), stacked under an element (`z-index`: 4), which results in a rendering order of `4-1`
    - ARTICLE #6: (`z-index`: 3), stacked under an element (`z-index`: 4), which results in a rendering order of `4-3`
    - ARTICLE #4: (`z-index`: 6), stacked under an element (`z-index`: 4), which results in a rendering order of `4-6`

>   - SECTION #1: (`z-index`: 5), which results in a rendering order of `5-0`

Reviewer: I missed this - this all needs to be updated to swap section elements to be the children of article element
